phoenix-core/src/main/java/org/apache/phoenix/mapreduce/VerifyReplicationTool.java

```diff
@@ -0,0 +1,477 @@
+    private Configuration conf;
+
+    private String zkQuorum;
+    private String tableName;
+    private String targetTableName;
+    private String targetZkQuorum;
+    private String sqlConditions;
+    private long timestamp;
+
+    VerifyReplicationTool(Configuration conf) {
+        this.conf = Preconditions.checkNotNull(conf, "Configuration cannot be null");
+    }
+
+    public static Builder newBuilder(Configuration conf) {
+        return new Builder(conf);
+    }
+
+    public static class Verifier
+            extends Mapper<NullWritable, VerifyReplicationSourceWritable, NullWritable, NullWritable> {
+
+        private QueryPlan targetQueryPlan;
+        private PhoenixResultSet targetResultSet = null;
+        private boolean targetHasNext;
+        private boolean sourceHasData;
+
+        public enum Counter {
+            GOODROWS, BADROWS, ONLY_IN_SOURCE_TABLE_ROWS, ONLY_IN_TARGET_TABLE_ROWS,
+            CONTENT_DIFFERENT_ROWS
+        }
+
+        @Override
+        protected void setup(Context context) throws IOException, InterruptedException {
+            super.setup(context);
+            try {
+                targetQueryPlan = PhoenixMapReduceUtil.getQueryPlan(context.getConfiguration(), true);
+            } catch (SQLException e) {
+                throw new IOException(e.getMessage());
+            }
+
+        }
+
+        @Override
+        protected void map(NullWritable key, VerifyReplicationSourceWritable value, Context context)
+                throws IOException, InterruptedException {
+
+            sourceHasData = true;
+            try {
+                if(targetResultSet == null) {
+                    Configuration conf = context.getConfiguration();
+                    byte[] targetStartRow = null;
+                    byte[] targetStopRow = null;
+                    // find source table split
+                    PhoenixInputSplit sourceInputSplit = (PhoenixInputSplit) context.getInputSplit();
+                    if(key != null) {
+                        targetStartRow = value.getSourceKey().get();
+                    }
+                    if(sourceInputSplit.getLength() != 0) {
+                        targetStopRow = sourceInputSplit.getKeyRange().getUpperRange();
+                    }
+                    getTargetResultSet(conf, targetStartRow, targetStopRow);
+                }
+
+                while(true) {
+                    if(!targetHasNext) {
+                        logFailRowAndIncrementCounter(context, Counter.ONLY_IN_SOURCE_TABLE_ROWS, value.getSourceKey());
+                        break;
+                    }
+                    ImmutableBytesWritable targetKey = null;
+                    if (targetResultSet.getCurrentRow() != null) {
+                        targetKey = new ImmutableBytesWritable();
+                        targetResultSet.getCurrentRow().getKey(targetKey);
+                    }
+
+                    int keyCompare;
+                    if(key == null){
```

Comment:
When can key be null?

Reply:
nit: Use ternary operator here

Reply:
Key cannot be null. It got left from a different approach I was testing earlier. I will remove it.